maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java

```diff
@@ -2846,8 +2849,96 @@ public void addProviderProperties()
         @Nonnull
         public Set<Artifact> getProviderClasspath()
         {
+            String provider = "surefire-junit-platform";
             String version = surefireBooterArtifact.getBaseVersion();
-            return dependencyResolver.getProviderClasspath( "surefire-junit-platform", version );
+            Set<Artifact> providerArtifacts = dependencyResolver.getProviderClasspath( provider, version );
+            alignJUnitPlatformVersion( providerArtifacts );
+            resolveJUnitJupiterEngine( providerArtifacts );
+            resolveJUnitVintageEngine( providerArtifacts );
+            return providerArtifacts;
+        }
+
+        private void resolveJUnitJupiterEngine( Set<Artifact> providerArtifacts )
+        {
+            Artifact junitJupiterApi = getProjectArtifactMap().get( "org.junit.jupiter:junit-jupiter-api" );
+            if ( junitJupiterApi == null ) // no api, no engine
+            {
+                return;
+            }
+            Artifact junitJupiterEngine = getProjectArtifactMap().get( "org.junit.jupiter:junit-jupiter-engine" );
+            if ( junitJupiterEngine != null ) // engine already resolved by project
+            {
+                return;
+            }
+            // resolve "junit-jupiter-engine" and its transitive dependencies
+            String jupiterVersion = junitJupiterApi.getBaseVersion();
+            resolve( providerArtifacts, "org.junit.jupiter", "junit-jupiter-engine", jupiterVersion );
+        }
+
+        private void resolveJUnitVintageEngine( Set<Artifact> providerArtifacts )
+        {
+            Artifact junit = getProjectArtifactMap().get( "junit:junit" );
+            if ( junit == null ) // no api, no engine
+            {
+                return;
+            }
+            if ( !junit.getBaseVersion().equals( "4.12" ) ) // not "JUnit 4.12", no engine
+            {
+                return;
+            }
+            Artifact junitVintageEngine = getProjectArtifactMap().get( "org.junit.vintage:junit-vintage-engine" );
+            if ( junitVintageEngine != null ) // engine already resolved by project
+            {
+                return;
+            }
+            // resolve "junit-vintage-engine" and its transitive dependencies
+            // heuristic: from Platform "x.y.z" to Vintage "5" + ".y.z"
```

Comment:
Who can say that the versions will be always like this?
Somebody has to watch the versions in Maven Central and still update this code. What if the version is not successfully resolved?

Reply:
> Who can say that the versions will be always like this?

The @junit-team -- I guess, it'll stay like this "heuristic" for the next 5-10 years.

> Somebody has to watch the versions in Maven Central and still update this code.

Sure. Unless we introduce a metadata engine artifact hint, like described here: https://github.com/junit-team/junit5/issues/1669

> What if the version is not successfully resolved?

The execution of Surefire fails for the project. Users may always provide the correct `junit-vintage-engine` in their project pom dependencies and prevent this code path to be executed.

Reply:
We are having low coverage without unit tests which is bad reputation for a test f/w. The green curve goes down but the red one is worse because uncovered lines grow with this commit since the lack of unit tests. The tests will be a good platform for making fast experiments rather than executing all the build and debugging a plugin. Unit testing must be a fun for a developer because he can see that it works - it's green again.

Reply:
@sormuras 
A motivation can be e.g. `AbstractSurefireMojoTest` or `AbstractSurefireMojoJava7PlusTest`.